elmslie-core/src/main/java/vivid/money/elmslie/core/switcher/Switcher.kt

```diff
@@ -25,47 +21,29 @@ import java.util.concurrent.TimeUnit
  */
 class Switcher {
 
-    private val disposable = CompositeDisposable()
-    private val tasks = mutableSetOf<ScheduledFuture<*>>()
-    private val service = ElmslieConfig.backgroundExecutor
+    @Volatile private var currentJob: Job? = null
+    private val switcherScope: CoroutineScope = CoroutineScope(ElmslieConfig.ioDispatchers)
 
     /**
-     * Executes [action] and cancels all previous requests scheduled on this [Switcher].
+     * Executes [action] as a job and cancels all previous ones.
      *
-     * @param delayMillis Operation delay measured with milliseconds.
-     *                    Can be specified to debounce existing requests.
-     * @param action New operation to be executed.
+     * @param delayMillis operation delay measured with milliseconds. Can be specified to debounce existing requests.
+     * @param action new operation to be executed.
      */
     fun switchInternal(
         delayMillis: Long = 0,
-        action: () -> Disposable,
-    ): Disposable {
-        disposable.clear()
-        synchronized(tasks) {
-            tasks.onEach { task -> task.cancel(true) }.clear()
-        }
-        val future = service.schedule(
-            { disposable.add(action()) },
-            delayMillis,
-            TimeUnit.MILLISECONDS
-        )
-        synchronized(tasks) {
-            tasks.add(future)
-        }
-        return Disposable { future.cancel(true) }
-    }
-
-    /**
-     * Awaits completion of all scheduled background tasks.
-     */
-    fun await(
-        delay: Long = 1,
-        timeUnit: TimeUnit = TimeUnit.DAYS
-    ) = tasks.forEach { task ->
-        try {
-            task.get(delay, timeUnit)
-        } catch (_: CancellationException) {
-            // Expected state
-        }
+        action: suspend () -> Unit,
+    ): Job? {
+        currentJob?.cancel()
+        return try {
+                switcherScope.launch {
+                    delay(delayMillis)
+                    action.invoke()
+                }
+            } catch (_: CancellationException) {
```

Comment:
CancellationException is coroutine specific Exception and will be consumed inside the coroutine, need to try/catch inside launch

Reply:
Don't get you(

Reply:
CancellationException can be thrown only inside coroutine, and launch\async etc will consume it automaticaly, so there is no way that we can catch it here

Reply:
here is a great video that you send to me early about coroutine interpreting - https://www.youtube.com/watch?v=neSHM6CpJvg